spring-data-mongodb/src/main/java/org/springframework/data/mongodb/util/BsonUtils.java

```diff
@@ -104,12 +123,26 @@ public static Map<String, Object> asMap(Bson bson) {
 	 * @since 3.2.5
 	 */
 	public static Document asDocument(Bson bson) {
+		return asDocument(bson, MongoClientSettings.getDefaultCodecRegistry());
+	}
 
-		if (bson instanceof Document) {
-			return (Document) bson;
+	/**
+	 * Return the {@link Bson} object as {@link Document}. Depending on the input type, the return value can be either a
+	 * casted version of {@code bson} or a converted (detached from the original value) using the given
+	 * {@link CodecRegistry} to obtain {@link org.bson.codecs.Codec codecs} that might be required for conversion.
+	 *
+	 * @param bson
+	 * @param codecRegistry must not be {@literal null}.
+	 * @return never {@literal null}.
+	 * @since 4.0
+	 */
+	public static Document asDocument(Bson bson, CodecRegistry codecRegistry) {
+
+		if (bson instanceof Document document) {
+			return document;
 		}
 
-		Map<String, Object> map = asMap(bson);
+		Map<String, Object> map = asMap(bson, codecRegistry);
```

Comment:
Given that `asMap` already does the `instanceof Document` check, this check may be safely removed here.